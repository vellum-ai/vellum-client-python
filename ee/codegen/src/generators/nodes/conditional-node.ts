import { python } from "@fern-api/python-ast";
import { Field } from "@fern-api/python-ast/Field";
import { Reference } from "@fern-api/python-ast/Reference";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { PORTS_CLASS_NAME } from "src/constants";
import { ConditionalNodeContext } from "src/context/node-context/conditional-node";
import { ConditionalNodePort } from "src/generators/conditional-node-port";
import { BaseSingleFileNode } from "src/generators/nodes/bases/single-file-base";
import {
  ConditionalNode as ConditionalNodeType,
  ConditionalNodeData,
  ConditionalRuleData,
} from "src/types/vellum";
import { isNilOrEmpty } from "src/utils/typing";

export class ConditionalNode extends BaseSingleFileNode<
  ConditionalNodeType,
  ConditionalNodeContext
> {
  protected getNodeClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    const inputFieldKeysByRuleId = new Map<string, string>();
    const valueInputKeysByRuleId = new Map<string, string>();
    this.constructMapContextForAllConditions(
      inputFieldKeysByRuleId,
      valueInputKeysByRuleId,
      this.nodeData.data
    );

    const baseNodeClassRef = this.getNodeBaseClass();

    const ref = python.reference({
      name: baseNodeClassRef.name,
      modulePath: baseNodeClassRef.modulePath,
      alias: baseNodeClassRef.alias,
      attribute: ["Ports"],
    });

    this.getNodeBaseClass().inheritReferences(ref);

    const portsClass = python.class_({
      name: "Ports",
      extends_: [ref],
    });
    Array.from(this.workflowContext.portContextById.entries()).forEach(
      ([portId, context]) => {
        const conditionDataWithIndex = [
          ...this.nodeData.data.conditions.entries(),
        ].find(([_, condition]) => condition.sourceHandleId === portId);

        if (!conditionDataWithIndex) {
          return;
        }

        portsClass.addField(
          python.field({
            name: context.portName,
            initializer: new ConditionalNodePort({
              portContext: context,
              inputFieldKeysByRuleId: inputFieldKeysByRuleId,
              valueInputKeysByRuleId: valueInputKeysByRuleId,
              conditionDataWithIndex: conditionDataWithIndex,
              nodeInputsByKey: this.nodeInputsByKey,
              nodeLabel: this.nodeData.data.label,
            }),
          })
        );
      }
    );

    statements.push(portsClass);
    return statements;
  }

  protected getNodeDisplayClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    statements.push(
      python.field({
        name: "label",
        initializer: python.TypeInstantiation.str(this.nodeData.data.label),
      }),
      python.field({
        name: "node_id",
        initializer: python.TypeInstantiation.uuid(this.nodeData.id),
      }),
      python.field({
        name: "target_handle_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.targetHandleId
        ),
      })
    );

    statements.push(
      python.field({
        name: "source_handle_ids",
        initializer: python.TypeInstantiation.dict(
          this.nodeData.data.conditions.map((condition, idx) => ({
            key: python.TypeInstantiation.int(idx),
            value: python.TypeInstantiation.uuid(condition.sourceHandleId),
          }))
        ),
      })
    );

    const ruleIdMapRef = python.reference({
      name: "RuleIdMap",
      modulePath: [
        ...this.workflowContext.sdkModulePathNames.NODE_DISPLAY_MODULE_PATH,
        "vellum",
        "conditional_node",
      ],
    });

    this.getNodeDisplayBaseClass().inheritReferences(ruleIdMapRef);

    statements.push(
      python.field({
        name: "rule_ids",
        initializer: python.TypeInstantiation.list(
          this.createRuleIdMapList(this.nodeData.data, ruleIdMapRef)
        ),
      })
    );

    const conditionIdRef = python.reference({
      name: "ConditionId",
      modulePath: [
        ...this.workflowContext.sdkModulePathNames.NODE_DISPLAY_MODULE_PATH,
        "vellum",
        "conditional_node",
      ],
    });

    this.getNodeDisplayBaseClass().inheritReferences(conditionIdRef);

    statements.push(
      python.field({
        name: "condition_ids",
        initializer: python.TypeInstantiation.list(
          this.createConditionIdList(this.nodeData.data, conditionIdRef)
        ),
      })
    );

    return statements;
  }

  protected getOutputDisplay(): Field | undefined {
    return undefined;
  }

  private createConditionIdList(
    nodeData: ConditionalNodeData,
    conditionIdRef: Reference
  ): AstNode[] {
    const conditionIdsList: AstNode[] = [];
    nodeData.conditions.forEach((condition) => {
      conditionIdsList.push(
        python.instantiateClass({
          classReference: conditionIdRef,
          arguments_: [
            python.methodArgument({
              name: "id",
              value: python.TypeInstantiation.uuid(condition.id),
            }),
            python.methodArgument({
              name: "rule_group_id",
              value: condition.data
                ? python.TypeInstantiation.uuid(condition.data.id)
                : python.TypeInstantiation.none(),
            }),
          ],
        })
      );
    });
    return conditionIdsList;
  }

  private createRuleIdMapList(
    nodeData: ConditionalNodeData,
    ruleIdMapRef: Reference
  ): AstNode[] {
    const ruleIdsList: AstNode[] = [];
    nodeData.conditions.forEach((condition) => {
      if (condition.data) {
        const ruleIdMap = this.createRuleIdMap(condition.data, ruleIdMapRef);
        if (ruleIdMap) {
          ruleIdsList.push(ruleIdMap);
        }
      }
    });

    return ruleIdsList;
  }

  private createRuleIdMap(
    ruleData: ConditionalRuleData,
    ruleIdMapRef: Reference
  ): AstNode | null {
    if (!ruleData) {
      return null;
    }

    let lhs = null;
    let rhs = null;
    let fieldId = null;
    let valueId = null;

    if (!ruleData.rules) {
      fieldId = ruleData.fieldNodeInputId;
      valueId = ruleData.valueNodeInputId;
    }

    // Check first rule in the arr (lhs)
    if (ruleData.rules && ruleData.rules[0]) {
      lhs = this.createRuleIdMap(ruleData.rules[0], ruleIdMapRef);
    }

    // Check second rule in the arr (rhs)
    if (ruleData.rules && ruleData.rules[1]) {
      rhs = this.createRuleIdMap(ruleData.rules[1], ruleIdMapRef);
    }

    return python.instantiateClass({
      classReference: ruleIdMapRef,
      arguments_: [
        python.methodArgument({
          name: "id",
          value: python.TypeInstantiation.uuid(ruleData.id),
        }),
        python.methodArgument({
          name: "lhs",
          value: lhs ? lhs : python.TypeInstantiation.none(),
        }),
        python.methodArgument({
          name: "rhs",
          value: rhs ? rhs : python.TypeInstantiation.none(),
        }),
        python.methodArgument({
          name: "field_node_input_id",
          value: fieldId
            ? python.TypeInstantiation.uuid(fieldId)
            : python.TypeInstantiation.none(),
        }),
        python.methodArgument({
          name: "value_node_input_id",
          value: valueId
            ? python.TypeInstantiation.uuid(valueId)
            : python.TypeInstantiation.none(),
        }),
      ],
    });
  }

  private constructMapContextForAllConditions(
    inputFieldKeysByRuleId: Map<string, string>,
    valueInputKeysByRuleId: Map<string, string>,
    nodeData: ConditionalNodeData
  ): void {
    nodeData.conditions.forEach((condition) => {
      if (condition.data) {
        this.constructMapContextByRuleIds(
          inputFieldKeysByRuleId,
          valueInputKeysByRuleId,
          condition.data
        );
      }
    });
  }

  private constructMapContextByRuleIds(
    inputFieldsByRuleId: Map<string, string>,
    valueInputsByRuleIds: Map<string, string>,
    ruleData: ConditionalRuleData
  ): void {
    if (!ruleData) {
      return;
    }
    if (isNilOrEmpty(ruleData.rules) && ruleData.fieldNodeInputId) {
      this.nodeData.inputs.forEach((input) => {
        if (input.id === ruleData.fieldNodeInputId) {
          inputFieldsByRuleId.set(ruleData.id, input.key);
        }
        if (
          ruleData.valueNodeInputId &&
          input.id === ruleData.valueNodeInputId
        ) {
          valueInputsByRuleIds.set(ruleData.id, input.key);
        }
      });
    }

    if (ruleData.rules) {
      ruleData.rules.forEach((rule) => {
        this.constructMapContextByRuleIds(
          inputFieldsByRuleId,
          valueInputsByRuleIds,
          rule
        );
      });
    }
  }

  protected getErrorOutputId(): undefined {
    return undefined;
  }

  protected getPortDisplay(): python.Field | undefined {
    const portDisplayOverridesDict = new Map();

    Array.from(this.workflowContext.portContextById.entries()).forEach(
      ([portId, context]) => {
        const conditionData = this.nodeData.data.conditions.find(
          (condition) => condition.sourceHandleId === portId
        );

        if (!conditionData) {
          return;
        }

        const edge = this.workflowContext.workflowRawEdges.find(
          (edge) => edge.sourceHandleId === portId
        );

        if (!edge) {
          return;
        }

        const portDisplayOverrides = python.instantiateClass({
          classReference: python.reference({
            name: "PortDisplayOverrides",
            modulePath:
              this.workflowContext.sdkModulePathNames
                .NODE_DISPLAY_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "id",
              value: python.TypeInstantiation.uuid(edge.sourceHandleId),
            }),
          ],
        });

        portDisplayOverridesDict.set(context.portName, portDisplayOverrides);
      }
    );
    return python.field({
      name: "port_displays",
      initializer: python.TypeInstantiation.dict(
        Array.from(portDisplayOverridesDict.entries()).map(([key, value]) => ({
          key: python.reference({
            name: this.nodeContext.nodeClassName,
            modulePath: this.nodeContext.nodeModulePath,
            attribute: [PORTS_CLASS_NAME, key],
          }),
          value: value,
        }))
      ),
    });
  }
}
