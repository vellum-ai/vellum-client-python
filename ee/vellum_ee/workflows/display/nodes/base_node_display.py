from functools import cached_property
import inspect
from uuid import UUID
from typing import TYPE_CHECKING, Any, Dict, Generic, Optional, Type, TypeVar, get_args

from vellum_ee.workflows.display.nodes.types import NodeOutputDisplay, PortDisplay, PortDisplayOverrides
from vellum_ee.workflows.display.utils.uuids import uuid4_from_hash
from vellum_ee.workflows.display.vellum import CodeResourceDefinition, NodeDefinition
from vellum.workflows.nodes.bases.base import BaseNode
from vellum.workflows.nodes.utils import get_wrapped_node, has_wrapped_node
from vellum.workflows.ports import Port
from vellum.workflows.references import OutputReference
from vellum.workflows.types.core import JsonObject
from vellum.workflows.types.generics import NodeType
from vellum.workflows.types.utils import get_original_base
from vellum.workflows.utils.names import pascal_to_title_case

if TYPE_CHECKING:
    from vellum_ee.workflows.display.types import WorkflowDisplayContext

_NodeDisplayAttrType = TypeVar("_NodeDisplayAttrType")


class BaseNodeDisplay(Generic[NodeType]):
    output_display: Dict[OutputReference, NodeOutputDisplay] = {}
    port_displays: Dict[Port, PortDisplayOverrides] = {}

    # Used to store the mapping between node types and their display classes
    _node_display_registry: Dict[Type[NodeType], Type["BaseNodeDisplay"]] = {}

    def __init__(self, node: Type[NodeType]):
        self._node = node

    def serialize(self, display_context: "WorkflowDisplayContext", **kwargs: Any) -> JsonObject:
        # raise NotImplementedError(f"Serialization for nodes of type {self._node.__name__} is not supported.")
        return {"type": "GENERIC"}

    def get_definition(self) -> NodeDefinition:
        node = self._node
        node_definition = NodeDefinition(
            name=node.__name__,
            module=node.__module__.split("."),
            bases=[
                CodeResourceDefinition(
                    name=base.__name__,
                    module=base.__module__.split("."),
                )
                for base in node.__bases__
            ],
        )
        return node_definition

    def get_node_output_display(self, output: OutputReference) -> NodeOutputDisplay:
        explicit_display = self.output_display.get(output)
        if explicit_display:
            return explicit_display

        return NodeOutputDisplay(id=uuid4_from_hash(f"{self.node_id}|{output.name}"), name=output.name)

    def get_node_port_display(self, port: Port) -> PortDisplay:
        overrides = self.port_displays.get(port)

        port_id: UUID
        if overrides:
            port_id = overrides.id
        else:
            port_id = uuid4_from_hash(f"{self.node_id}|ports|{port.name}")

        return PortDisplay(id=port_id, node_id=self.node_id)

    @classmethod
    def get_from_node_display_registry(cls, node_class: Type[NodeType]) -> Type["BaseNodeDisplay"]:
        return cls._node_display_registry[node_class]

    @cached_property
    def node_id(self) -> UUID:
        """Can be overridden as a class attribute to specify a custom node id."""
        return uuid4_from_hash(self._node.__qualname__)

    @cached_property
    def label(self) -> str:
        """Can be overridden as a class attribute to specify a custom label."""
        return pascal_to_title_case(self._node.__name__)

    @classmethod
    def _get_explicit_node_display_attr(
        cls,
        attribute: str,
        attribute_type: Type[_NodeDisplayAttrType],
    ) -> Optional[_NodeDisplayAttrType]:
        node_display_attribute: Optional[_NodeDisplayAttrType] = getattr(cls, attribute, None)

        if node_display_attribute is None:
            return None

        if isinstance(node_display_attribute, attribute_type):
            return node_display_attribute

        raise ValueError(f"Node {cls.__name__} must define an explicit {attribute} of type {attribute_type.__name__}.")

    def __init_subclass__(cls, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)

        original_base = get_original_base(cls)
        node_class = get_args(original_base)[0]
        if isinstance(node_class, TypeVar):
            bounded_class = node_class.__bound__
            if inspect.isclass(bounded_class) and issubclass(bounded_class, BaseNode):
                cls._node_display_registry[bounded_class] = cls
        elif issubclass(node_class, BaseNode):
            if has_wrapped_node(node_class):
                wrapped_node = get_wrapped_node(node_class)
                if wrapped_node._is_wrapped_node:
                    cls._node_display_registry[wrapped_node] = cls
                    return

            cls._node_display_registry[node_class] = cls
