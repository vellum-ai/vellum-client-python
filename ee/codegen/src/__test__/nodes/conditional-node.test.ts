import { Writer } from "@fern-api/python-ast/core/Writer";
import { beforeEach } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import {
  conditionalNodeFactory,
  conditionalNodeWithNullOperatorFactory,
} from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { ConditionalNodeContext } from "src/context/node-context/conditional-node";
import { ConditionalNode } from "src/generators/nodes/conditional-node";
import {
  ConditionalNode as ConditionalNodeType,
  WorkflowNodeType,
} from "src/types/vellum";

describe("ConditionalNode", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const nodeData = conditionalNodeFactory({ includeElif: true });

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;
    workflowContext.addNodeContext(nodeContext);

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile", async () => {
    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDisplayFile", async () => {
    node.getNodeDisplayFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDefinition", () => {
    expect(node.nodeContext.getNodeDefinition()).toMatchSnapshot();
  });
});

describe("ConditionalNode with null operator", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const nodeData = conditionalNodeWithNullOperatorFactory();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;
    workflowContext.addNodeContext(nodeContext);

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile", async () => {
    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDisplayFile", async () => {
    node.getNodeDisplayFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDefinition", () => {
    expect(node.nodeContext.getNodeDefinition()).toMatchSnapshot();
  });
});

describe("ConditionalNode with incorrect rule id references", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const invalidNodeData = constructInvalidConditionalNode();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData: invalidNodeData,
    })) as ConditionalNodeContext;
    workflowContext.addNodeContext(nodeContext);

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile should throw error", async () => {
    try {
      node.getNodeFile().write(writer);
      await writer.toStringFormatted();
    } catch (error: unknown) {
      if (error instanceof Error) {
        expect(error.message).toBe(
          "Could not find input field key given ruleId: ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc on rule index: 0 on condition index: 0 for node: Conditional Node"
        );
      } else {
        throw new Error("Unexpected error type");
      }
    }
  });

  function constructInvalidConditionalNode(): ConditionalNodeType {
    return {
      id: "b81a4453-7b80-41ea-bd55-c62df8878fd3",
      type: WorkflowNodeType.CONDITIONAL,
      data: {
        label: "Conditional Node",
        targetHandleId: "842b9dda-7977-47ad-a322-eb15b4c7069d",
        conditions: [
          {
            id: "8d0d8b56-6c17-4684-9f16-45dd6ce23060",
            type: "IF",
            sourceHandleId: "63345ab5-1a4d-48a1-ad33-91bec41f92a5",
            data: {
              id: "fa50fb0c-8d62-40e3-bd88-080b52efd4b2",
              rules: [
                {
                  id: "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc",
                  rules: [],
                  fieldNodeInputId: "2cb6582e-c329-4952-8598-097830b766c7",
                  operator: "=",
                  valueNodeInputId: "cf63d0ad-5e52-4031-a29f-922e7004cdd8",
                },
              ],
              combinator: "AND",
            },
          },
          {
            id: "ea63ccd5-3fe3-4371-ba3c-6d3ec7ca2b60",
            type: "ELSE",
            sourceHandleId: "14a8b603-6039-4491-92d4-868a4dae4c15",
          },
        ],
        version: "2",
      },
      inputs: [
        {
          id: "non-existent-id",
          key: "non-existent-rule-id.field",
          value: {
            rules: [
              {
                type: "INPUT_VARIABLE",
                data: {
                  inputVariableId: "d2287fee-98fb-421c-9464-e54d8f70f046",
                },
              },
            ],
            combinator: "OR",
          },
        },
      ],
      displayData: {
        width: 480,
        height: 180,
        position: {
          x: 2247.2797390213086,
          y: 30.917121251477084,
        },
      },
    };
  }
});
