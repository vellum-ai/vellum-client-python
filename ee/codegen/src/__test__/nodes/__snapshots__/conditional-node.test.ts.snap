// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConditionalNode > getNodeDefinition 1`] = `
{
  "bases": [
    {
      "module": [
        "vellum",
        "workflows",
        "nodes",
        "displayable",
      ],
      "name": "ConditionalNode",
    },
  ],
  "module": [
    "code",
    "nodes",
    "conditional_node",
  ],
  "name": "ConditionalNode",
}
`;

exports[`ConditionalNode > getNodeDisplayFile 1`] = `
"from vellum_ee.workflows.display.nodes import BaseConditionalNodeDisplay
from ...nodes.conditional_node import ConditionalNode
from uuid import UUID
from vellum_ee.workflows.display.nodes.vellum.conditional_node import (
    RuleIdMap,
    ConditionId,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition


class ConditionalNodeDisplay(BaseConditionalNodeDisplay[ConditionalNode]):
    label = "Conditional Node"
    node_id = UUID("b81a4453-7b80-41ea-bd55-c62df8878fd3")
    target_handle_id = UUID("842b9dda-7977-47ad-a322-eb15b4c7069d")
    source_handle_ids = {
        0: UUID("63345ab5-1a4d-48a1-ad33-91bec41f92a5"),
        1: UUID("2c03f27f-ea64-42fc-8a6c-383550c58ae4"),
        2: UUID("14a8b603-6039-4491-92d4-868a4dae4c15"),
    }
    rule_ids = [
        RuleIdMap(
            id=UUID("fa50fb0c-8d62-40e3-bd88-080b52efd4b2"),
            lhs=RuleIdMap(
                id=UUID("ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc"),
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            rhs=None,
            field_node_input_id=None,
            value_node_input_id=None,
        ),
        RuleIdMap(
            id=UUID("fa50fb0c-8d62-40e3-bd88-080b52efd4b2"),
            lhs=RuleIdMap(
                id=UUID("ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc"),
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            rhs=None,
            field_node_input_id=None,
            value_node_input_id=None,
        ),
    ]
    condition_ids = [
        ConditionId(
            id=UUID("8d0d8b56-6c17-4684-9f16-45dd6ce23060"),
            rule_group_id=UUID("fa50fb0c-8d62-40e3-bd88-080b52efd4b2"),
        ),
        ConditionId(
            id=UUID("e63c3933-ef86-451f-88bc-d7ea7dce4310"),
            rule_group_id=UUID("fa50fb0c-8d62-40e3-bd88-080b52efd4b2"),
        ),
        ConditionId(
            id=UUID("ea63ccd5-3fe3-4371-ba3c-6d3ec7ca2b60"), rule_group_id=None
        ),
    ]
    node_input_ids_by_name = {
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.field": UUID(
            "2cb6582e-c329-4952-8598-097830b766c7"
        ),
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.value": UUID(
            "cf63d0ad-5e52-4031-a29f-922e7004cdd8"
        ),
    }
    port_displays = {}
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2247.2797390213086, y=30.917121251477084),
        width=480,
        height=180,
    )
"
`;

exports[`ConditionalNode > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ConditionalNode as BaseConditionalNode
from vellum.workflows.ports import Port
from ..inputs import Inputs


class ConditionalNode(BaseConditionalNode):
    class Ports(BaseConditionalNode.Ports):
        branch_1 = Port.on_if(Inputs.field.equals("testtest"))
        branch_2 = Port.on_elif(Inputs.field.is_not_null())
        branch_3 = Port.on_else()
"
`;

exports[`ConditionalNode with null operator > getNodeDefinition 1`] = `
{
  "bases": [
    {
      "module": [
        "vellum",
        "workflows",
        "nodes",
        "displayable",
      ],
      "name": "ConditionalNode",
    },
  ],
  "module": [
    "code",
    "nodes",
    "conditional_node",
  ],
  "name": "ConditionalNode",
}
`;

exports[`ConditionalNode with null operator > getNodeDisplayFile 1`] = `
"from vellum_ee.workflows.display.nodes import BaseConditionalNodeDisplay
from ...nodes.conditional_node import ConditionalNode
from uuid import UUID
from vellum_ee.workflows.display.nodes.vellum.conditional_node import (
    RuleIdMap,
    ConditionId,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition


class ConditionalNodeDisplay(BaseConditionalNodeDisplay[ConditionalNode]):
    label = "Conditional Node"
    node_id = UUID("b81a4453-7b80-41ea-bd55-c62df8878fd3")
    target_handle_id = UUID("842b9dda-7977-47ad-a322-eb15b4c7069d")
    source_handle_ids = {0: UUID("63345ab5-1a4d-48a1-ad33-91bec41f92a5")}
    rule_ids = [
        RuleIdMap(
            id=UUID("fa50fb0c-8d62-40e3-bd88-080b52efd4b2"),
            lhs=RuleIdMap(
                id=UUID("ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc"),
                lhs=None,
                rhs=None,
                field_node_input_id=None,
                value_node_input_id=None,
            ),
            rhs=None,
            field_node_input_id=None,
            value_node_input_id=None,
        )
    ]
    condition_ids = [
        ConditionId(
            id=UUID("8d0d8b56-6c17-4684-9f16-45dd6ce23060"),
            rule_group_id=UUID("fa50fb0c-8d62-40e3-bd88-080b52efd4b2"),
        )
    ]
    node_input_ids_by_name = {
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.field": UUID(
            "2cb6582e-c329-4952-8598-097830b766c7"
        )
    }
    port_displays = {}
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2247.2797390213086, y=30.917121251477084),
        width=480,
        height=180,
    )
"
`;

exports[`ConditionalNode with null operator > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ConditionalNode as BaseConditionalNode
from vellum.workflows.ports import Port
from ..inputs import Inputs


class ConditionalNode(BaseConditionalNode):
    class Ports(BaseConditionalNode.Ports):
        branch_1 = Port.on_if(Inputs.field.is_null())
"
`;
